modules/runner/src/integrationTest/java/org/apache/ignite/internal/raftsnapshot/ItTableRaftSnapshotsTest.java

```diff
@@ -0,0 +1,781 @@
+                Thread.currentThread().interrupt();
+
+                throw new RuntimeException(e);
+            } catch (ExecutionException | TimeoutException e) {
+                throw new RuntimeException(e);
+            }
+
+            nodes.set(index, newIgniteNode);
+        }
+
+        RaftGroupService leaderServiceFor(TablePartitionId tablePartitionId) throws InterruptedException {
+            AtomicReference<RaftGroupService> serviceRef = new AtomicReference<>();
+
+            assertTrue(
+                    waitForCondition(() -> {
+                        RaftGroupService service = currentLeaderServiceFor(tablePartitionId);
+
+                        serviceRef.set(service);
+
+                        return service != null;
+                    }, 10_000),
+                    "Did not find a leader for " + tablePartitionId + " in time"
+            );
+
+            RaftGroupService result = serviceRef.get();
+
+            assertNotNull(result);
+
+            return result;
+        }
+
+        @Nullable
+        private RaftGroupService currentLeaderServiceFor(TablePartitionId tablePartitionId) {
+            return aliveNodes()
+                    .map(IgniteImpl.class::cast)
+                    .map(ignite -> {
+                        JraftServerImpl server = (JraftServerImpl) ignite.raftManager().server();
+
+                        Optional<RaftNodeId> maybeRaftNodeId = server.localNodes().stream()
+                                .filter(nodeId -> nodeId.groupId().equals(tablePartitionId))
+                                .findAny();
+
+                        return maybeRaftNodeId.map(server::raftGroupService).orElse(null);
+                    })
+                    .filter(Objects::nonNull)
+                    .filter(service -> service.getRaftNode().isLeader())
+                    .findAny()
+                    .orElse(null);
+        }
+
+        private Stream<IgniteImpl> aliveNodes() {
+            return nodes.stream().filter(Objects::nonNull);
+        }
+
+        private Session openSession() {
+            return entryNode().sql()
+                    .sessionBuilder()
+                    .defaultSchema("PUBLIC")
+                    .defaultQueryTimeout(QUERY_TIMEOUT_MS, TimeUnit.MILLISECONDS)
+                    .build();
+        }
+
+        void shutdown() {
+            aliveNodes().forEach(node -> IgnitionManager.stop(node.name()));
+        }
+
+        private void knockOutNode(int nodeIndex, NodeKnockout knockout) {
+            knockout.knockOutNode(nodeIndex, this);
+
+            knockedOutIndices.add(nodeIndex);
+        }
+
+        private void reanimateNode(int nodeIndex, NodeKnockout knockout) {
+            knockout.reanimateNode(nodeIndex, this);
+
+            knockedOutIndices.remove(nodeIndex);
+        }
+    }
+
+    /**
+     * A way to make a node be separated from a cluster and stop receiving updates.
+     */
+    private enum NodeKnockout {
```

Comment:
Why enum and not just methods and functions?

Reply:
To make it usable with `@ParameterizedTest` via `@EnumSource`